### DISCUSSION TITLE
Add support for monitoring probes in CGNS

### DISCUSSION THREAD

**Discussion by MicK7** [CGNS/CGNS#339] [CGNStalk]
Hello,

here is a new CPEX proposal: https://cgnsorg.atlassian.net/browse/CGNS-272

The goal is to add a generic support for numerical probes in the CGNS structures.
Thus retrieving physical monitoring data for unsteady flow or during convergence process will be easier.

Please feel free to comment here. Any good idea is welcomed.

**Comment by mfolusiak** [comment]
Hi, I remember we had discussed similar idea on CGNSTalk group ('Data series of scalar data' thread). What I proposed and implemented in production in the code (FLACSCFD) is very similar:
```
  +-Results1d  -- CGNSBase_t I4 (2)
  | +-BaseIterativeData  -- BaseIterativeData_t I4 (1)
  | | +-TimeValues  -- DataArray_t R4 (300)
  | | +-Timestep  -- DataArray_t R4 (300)
  | | +-IterationValues  -- DataArray_t I4 (300)
  | +-Monitors  -- UserDefinedData_t MT ()
  | | +-Points  -- UserDefinedData_t I4 (1)
  | | | +-Point1  -- UserDefinedData_t MT ()
  | | | | +-Pressure  -- DataArray_t R4 (300)
```
My idea never got to the CPEX stage so I am glad that you @MicK7  took an attempt and I like very much what you proposed. I would like to share some experience and feedback from users who are using our software (with this structure implemented for several years now). Users run on both desktops and HPCs. The data in my code is flushed every say 10 minutes and simulations can run for minutes, hours or as long as 2-4 weeks.
I would like to discuss flaws of the HDF5 files and how to address them or minimize the risk by adjusting the structure of the CGNS:
* single file is not written incrementally - small change in the file , especially rewriting `DataArray_t` changes the structure drastically. This affects the performance of incremental downloads (e.g. rsync). Most our users prefer to download results incrementally. In case of 3D data it is easy ot address that by grouping zones and creating new file for each consecutive group and then linking these zones back to the main CGNS file of a minimal size, containing grid and iterative data.
* when HDF5 file gets corrupted it is close to impossible to recover the data. All data in the file are lost. The file can very easily get corrupted. Examples:
  * open your file in cgnsview and hit Ctrl+C - even simple viewer can destroy the file!
  * disk runs out of space while writing

From the above flaws - requirements for good design for probes and any time of timeseries:
* developer should be able to split timeseries into chunks and put them in linked files. Right now neither my design nor the proposed design facilitates this. One simple solution is to create multiple `Base_t` nodes, each one with data for range of iterations.
```
  +-Results1d_#1-100  -- CGNSBase_t I4 (2)
  +-Results1d_#101-200  -- CGNSBase_t I4 (2)
  +-Results1d_#201-300  -- CGNSBase_t I4 (2)
```
However the proposed solution is not self-contained, there is no structure describing the sequence in which data in bases should be glued together - a bad idea.
* I believe a better solution would be to only have one `Base_t` for Probe(1D) data and separate from the 3D base. The information about the sequence of the data should be written to the `BaseIterativeData_t` node in some way. Again the easiest, but very inefficient solution is to have a string pointer for each timestep in node `MonitrorsPointers` - in a similar way we create pointers in `ZonePointers`. I don't have an idea for a better solution that would be self-contained, efficient and follow the design philosophy of CGNS. Any thoughts?
```
  +-Results1d  -- CGNSBase_t I4 (2)
  | +-BaseIterativeData  -- BaseIterativeData_t I4 (1)
  | | +-TimeValues  -- DataArray_t R4 (300)
  | | +-Timestep  -- DataArray_t R4 (300)
  | | +-IterationValues  -- DataArray_t I4 (300)
  | | +-IterationValues  -- DataArray_t I4 (300)
  | | +-MonitrorsPointers (????)
  | +-Monitors_#1-100  -- UserDefinedData_t MT ()
  | | +-Points  -- UserDefinedData_t I4 (1)
  | | | +-Point1  -- UserDefinedData_t MT ()
  | | | | +-Pressure  -- DataArray_t R4 (100)
  | +-Monitors_#101-200  -- UserDefinedData_t MT ()
  | | +-Points  -- UserDefinedData_t I4 (1)
  | | | +-Point1  -- UserDefinedData_t MT ()
  | | | | +-Pressure  -- DataArray_t R4 (100)
  | +-Monitors_#201-300  -- UserDefinedData_t MT ()
  | | +-Points  -- UserDefinedData_t I4 (1)
  | | | +-Point1  -- UserDefinedData_t MT ()
  | | | | +-Pressure  -- DataArray_t R4 (100)
```

**Reply by MicK7** [reply]
Thanks for your detailed feedback.

To give more information on the proposal, the idea was that for DataArray_t located in Probes they will be build as hdf5 chunked dataset. Thus the developer will have a function to create the array and maybe set its prefered chunked size. Then an "append" like function should extend the array or else the developer should reserve the appropriate size at the beginning. As it is an hdf5 chunked array I hope it helps with corruption.
@brtnfld Do you have any opinion ?

As you propose, we can think of letting the developper chunck by hand the written data and add an "iteration range" attribute to not depend on a MonitorsPointers node.

**Reply by brtnfld** [reply]
Chunking will not help with corruption. To address file corruption, one can manage the metadata cache:
 https://support.hdfgroup.org/HDF5/docNewFeatures/FineTuneMDC/RFC%20H5Ocork%20v5%20new%20fxn%20names.pdf
If it is essential to inspect an active HDF5 file, then SWMR (single writer, multiple readers) could be an alternative option.

Flushing the data after each write might severely impact performance, but it is the way to avoid file corruption. Alternatively, one can use an external raw data file and link those files in the CGNS/HDF5 file.
